### ISSUE TITLE
Error if a folder/label is missing

### ISSUE THREAD

**Issue by pprw** [OfflineIMAP/offlineimap#502]
bug: **Offlineimap get stuck if I delete a folder/label from the gmail interface.**

#### General informations

- system/distribution (with version):  Debian sid up-to-date
- offlineimap version (`offlineimap -V`):  offlineimap v7.1.3, imaplib2 v2.57 (system), Python v2.7.14, OpenSSL 1.1.0f  25 May 2017
- Python version: see below
- server name or domain: Gmail
- CLI options:  none

#### Configuration file offlineimaprc

```
[general]
ui = ttyui
accounts = Petur
maxsyncaccounts = 1

[mbnames]
enabled = yes
filename = ~/.mutt/mailboxes
header = "mailboxes "
peritem = "+%(accountname)s/%(foldername)s"
sep = " "
footer = "\n"

[Account Petur]
localrepository = Petur-Local
remoterepository = Petur-Remote

synclabels = yes
autorefresh = 0.5
quick = 10


[Repository Petur-Local]
type = GmailMaildir
localfolders = ~/Mail/Petur


[Repository Petur-Remote]
type = Gmail
remoteuser = my_email
remotepass = secret
maxconnections = 3
sslcacertfile = /etc/ssl/certs/ca-certificates.crt

keepalive = 60
holdconnectionopen = yes
```

#### Logs, error

```
$ offlineimap
OfflineIMAP 7.1.3
  Licensed under the GNU GPL v2 or any later version (with an OpenSSL exception)
imaplib2 v2.57 (system), Python v2.7.14, OpenSSL 1.1.0f  25 May 2017
Account sync Petur:
 *** Processing account Petur
 Establishing connection to imap.gmail.com:993 (Petur-Remote)
 ERROR: While attempting to sync account 'Petur'
  getfolder() got an unexpected keyword argument 'decode'
 *** Finished account 'Petur' in 0:01

```

#### Steps to reproduce the error

- Launch offlineimap once to create folders and download emails.
- Delete one folder/label from Gmail interface
- start again offlineimap which will generate the previously pasted error.

**Comment by nicolas33** [comment]
I've sketched a patch that should fix this issue. Could you please try the above version?

**Comment by pprw** [comment]
Seems to work!

**Comment by nicolas33** [comment]
Thanks for your feedback.

@uliska Wanna review the patch before I merge?

**Comment by uliska** [comment]
As I've stated clearly I did *not* look into the GMail code and asked for specific testing.
That said, your patch looks like a necessary complement/completion.

**Comment by iliastsi** [comment]
@nicolas33 I do not understand why your patch works. Shouldn't you also add `decode=True` on GmailFolder's init method (since self.getfoldertype() returns folder.Gmail.GmailFolder) or am I missing something?

**Comment by nicolas33** [comment]
@uliska The issue is that the utf8foldernames series introduce regression so I thought you might like to review additions to your job. Sorry if this bothers you.

@iliastsi You're right. I've sketched this patch quickly yesterday night. Now that I look at it more deeply I'm stuck at why this works... I'm still trying to figure this out...

**Comment by nicolas33** [comment]
@pprw Is there any chance that you missed any error while testing the patch?

**Comment by nicolas33** [comment]
Here is a new version of the patch.

**Comment by uliska** [comment]
Am 27.10.2017 um 18:01 schrieb Nicolas Sebrecht:
> @uliska <https://github.com/uliska> The issue is that the 
> utf8foldernames series introduce regression so I thought you might 
> like to review additions to your job. Sorry if this bothers you.

No, it doesn't bother me. I was aware that my phrasing could be 
misunderstood but didn't find a better way to put it. What I wanted to 
say is that I don't have any specific insight in any possible GMail code 
related problems.

**Comment by nicolas33** [comment]
> No, it doesn't bother me. I was aware that my phrasing could be
> misunderstood but didn't find a better way to put it. What I wanted to
> say is that I don't have any specific insight in any possible GMail code
> related problems

Ok! Thanks!

**Comment by pprw** [comment]
It is totally possible I missed errors. After I applied the first patch, emails appears in mutt (offlineimap was in automatic mode). I assumed the problem was fixed but didn't look for errors. It is also possible the emails were downloaded for another reason (I was heavily working on optimizing  my mails, making a lot of changes in folders in particular). Sorry for that.

**Comment by nicolas33** [comment]
Ok. Could you try the latest patch, please?

**Comment by pprw** [comment]
Ok so:
1. I created a label in my gmail account and launch offlineimap. The label is synchronized and a folder is created.
2. I delete the label from gmail interface. I launch offlineimap and I have the error of post #1.
3. I modified the three files of the patch fix (I want to stay with the debian stable repo version and not use git clone).
4. Now I have this error when launching offlineimap:
(The config is slightly different from the one in post #1, I have two accounts (Gmail and Petur) and I did the test "create and delete label" on the gmail account.)
```
$ offlineimap -o
OfflineIMAP 7.1.3
  Licensed under the GNU GPL v2 or any later version (with an OpenSSL exception)
ERROR: Exceptions occurred during the run!
ERROR: While attempting to sync account 'Gmail
  __init__() takes exactly 5 arguments (4 given)

Traceback:
  File "/usr/share/offlineimap/offlineimap/accounts.py", line 283, in syncrunner
    self.__sync()
  File "/usr/share/offlineimap/offlineimap/accounts.py", line 359, in __sync
    remoterepos.getfolders()
  File "/usr/share/offlineimap/offlineimap/repository/IMAP.py", line 484, in getfolders
    self))

ERROR: While attempting to sync account 'Petur'
  __init__() takes exactly 5 arguments (4 given)

Traceback:
  File "/usr/share/offlineimap/offlineimap/accounts.py", line 283, in syncrunner
    self.__sync()
  File "/usr/share/offlineimap/offlineimap/accounts.py", line 359, in __sync
    remoterepos.getfolders()
  File "/usr/share/offlineimap/offlineimap/repository/IMAP.py", line 484, in getfolders
    self))

```

**Comment by nicolas33** [comment]
Thank you for the detailed feedback. Very much appreciated.
You're right we have two use cases. Both must run.

I sucked! I forgot one remaining call so here is a new patch.
I expect it to be the correct and last one. Could you please try it?

**Comment by nicolas33** [comment]
Wait, it's broken.

**Comment by nicolas33** [comment]
@uliska I'm trying to update the changes about "decode" for Gmail. My latest patch above introduces regressions for non-Gmail accounts compared to yours (aka: the current 'next' branch) but I don't get why.

Could you help me to understand why, please?

**Comment by uliska** [comment]
I'll try to find the time.
-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

**Comment by nicolas33** [comment]
@uliska I think I've found my mistake. I was setting `decode=False` line 483 in `repository/IMAP.py`.

BTW, I would expect to give UTF-8 foldernames to the `--delete-folder` CLI option. This does not work for now. I'll check this out in #505.

@pprw Could you please try this latest patch above?

I'll release v7.1.4 once this bug report gets closed.

**Comment by pprw** [comment]
Ok I tried the last patch and everything is working for my. I cannot test with non-gmail account since my two account are gmail's ones.

**Comment by nicolas33** [comment]
Ok, thank you much all.
I've released v7.1.4 just now. ,-)

Have fun!
